README.md
# MPI-optimised version of RSienaTest

Original source from https://rdrr.io/rforge/RSienaTest/, based on version 1.2-30 (July 13, 2021).

There is an install_github function to install R packages hosted on GitHub in the `devtools` package.

```
install_github("ADACS-Australia/RSienaTest")
```

Needs [mpi-Rscript](https://github.com/ADACS-Australia/mpi-Rscript) to run properly.

Example usage:
```
mpirun -n 4 mpi-Rscript example.R
```

example.R:
```
library(RSienaTest)

# load data and initial estimation
load('example.RData')

groupModel.ec <- sienaBayes(GroupsModel, data = example,
                            effects = GroupEffects, priorMu = Mu, priorSigma = Sig,
                            priorKappa = 0.01,
                            prevBayes = groupModel.e,
                            nmain=3, nrunMHBatches=40,
                            silentstart=FALSE,
                            clusterType="MPI")
```


ChangeLog
2021-07-13 R-Forge Revision 354, package version 1.2-30.
Changes in RSienaTest:
   * sienaBayes: adaptation of nwarm and nmain 
     at the end of initializeBayes dropped.

2021-04-14 R-Forge Revision 353, package version 1.2-29.
Changes in RSienaTest:
   * sienaBayes: break out of prewarming if in 2 consecutive steps
     there are no acceptances.
   * sienaBayes: if nrunMHBatches is odd, 1 is added to it.
   * ThinBayesAcceptances mentioned in help page for sienaBayes.
   * Test16 omitted.

2021-03-18 R-Forge Revision 352, package version 1.2-28.
Changes in RSienaTest:
   * New effect homXTransRecTrip
   * Added rateParameterPosition	to glueBayes, which results
     in correct effect names for extract.posteriorMeans.

2020-09-16 R-Forge Revision 350, package version 1.2-27.
Changes in RSiena and RSienaTest:
   * Changed requirement for tcltk to "Suggests", 
     and further modified / cleaned up DESCRIPTION.
   * In siena07: if (!requireNamespace("tcltk")) set batch to TRUE.
   * New effect transTripX.
   * For effect from.w.ind, option parameter=-1 added.
   * The to effect is an ego effect.
   * For siena.table, some of the effectNames changed to nice strings,
     so that LaTeX can run without errors if type='tex'.
   * The object produced by siena08 now has IWLS estimates more easily 
     accessible, as object$muhat and object$se.muhat.
   * Error message in sienaTimeTest for sienaFit objects produced with
     lessMem=TRUE.
   * More extensive error message for error in named vectors in algorithm object
     (checkNames in initializeFRAN).
   * For sienaDataCreate: more extensive error message, and class(...) replaced
     by class(...)[1]. 
   * multiplication factor added to print.sienaAlgorithm if maxlike.
   * SienaAlgorithmCreate: requirements for mult corrected in help page.
   * In sienaAlgorithmCreate, use the definitions for projname=NULL
     also if any environment variable _R_CHECK* is set.
Changes in RSiena:
   * Vignette basicRSiena added (was earlier available as a script);
     thanks to James Hollway.
Changes in RSienaTest:
   * Improved initialisation in sienaBayes: initial scale factors (squared)
     and initial parameter estimate ('Kelley').
   * Allow parameter targetMHProb of sienaBayes to have length 2.
   * Test on class()== replaced by inherits() in sienaBayes.
   * Another attempt to discard the objects zn and zsmall remaining after operation
     of sienaBayes.


2020-04-08 R-Forge Revision 348, package version 1.2-25.
Changes in RSiena and RSienaTest:
   * Correction of sienaDataCreate for actor covariates (gave a warning).
   * Auxiliary function dyadicCov now can handle missings,
     does not return a named value for frequency of zeros,
     and returns the frequencies without any division.
   * descriptives.sienaGOF also shows the exceedance probabilities
     of the observed statistics.
   * modelType and behModelType, if NULL, get their defaults already in
     initializeFRAN (checkNames) rather than in C++.
   * Some extra information in help page for siena07.
Changes in RSienaTest:
   * getNames corrected for data with multiple periods
     (used in sienaBayes, extract.sienaBayes, and extract.posteriorMeans)
  

2020-02-16 R-Forge Revision 347, package version 1.2-24.
Changes in RSiena and RSienaTest:
   * New auxiliary function dyadicCov.
   * Correction of error in modelType when one-mode as well as two-mode
     networks are used (initializeFRAN.r).
Changes in RSienaTest:
   * Corrected siena.table for sienaBayesFit objects;
     added startingDate in sienaBayes and siena08 to the object produced. 

2020-01-15 R-Forge Revision 346, version 1.2-23 same as CRAN version
(except that one error in dontrun part of sienaGOF-auxiliary.Rd was corrected).

2020-01-12 Package version 1.2-23 submitted to CRAN.
   * Changes in RSiena and RSienaTest:
     Option projname=NULL in sienaAlgorithmCreate.
     This option is used for all examples in RSiena where siena07 is called,
     to restrict file writing for examples to the temporary directory. 

2020-01-10 Package version 1.2-22.
   * Changes in RSiena and RSienaTest:
     Changed extension names of output files from .out to .txt. 

2020-01-04 R-Forge Revision 345, package version 1.2-22.
Changes in RSiena and RSienaTest:
   * Tested effects reported in print of Multipar.RSiena and score.Test.
   * Reordered effects object so that reciAct comes after instead of 
     before inAct and inAct.c.
   * Operation of siena07 in Test 14 silenced (message Brian Ripley about 
     apparently minor changes depending on the platform).

2019-12-18 packages version 1.2-21.
Changes in RSiena and RSienaTest:
   * Tests ended with deleting file Siena.out by unlink instead of file.remove.

2019-12-16 packages version 1.2-20.
Changes in RSiena:
   * dumpChain (Chain) and getChainProbabilities (siena07models) dropped; 
     kept in RSienaTest.
   * Configure line 2274 changed to RBIN="${R_HOME}/bin"
Changes in RSiena and RSienaTest:
   * Tests ended with deleting file Siena.out.

2019-12-16 R-Forge Revision 344, package version 1.2-19.
Changes in RSiena and RSienaTest:
   * SienaTimeTest: some corrections with consequences probably only if
     one effect is being tested; or is left after automatic exclusion
     of tested effects.
   * Setting-related effects corrected.
   * Some further changes in Settings model.
   * New effects avGroupEgoX, outIso.
   * CovariateDependentNetworkEffect:  + lSimulatedOffset taken out of
     definition missings for ChangingCovariate (seems wrong).
   * Alphabetically ordered sources.list.
Changes in RSienaTest:
   * In sienaBayes: more precise check that 
     prevAns has same specification as effects;
     priorSigEta added to sienaBayesFit object.
	  dimnames(ThinParameters)[[3]] defined as effect names;

R/siena07.r
##/*****************************************************************************
## * SIENA: Simulation Investigation for Empirical Network Analysis
## *
## * Web: http://www.stats.ox.ac.uk/~snijders/siena
## *
## * File: siena07.r
## *
## * Description: This file contains the main controlling module for the
## * estimation.
## * Also contains utility functions used within siena07
## ****************************************************************************/

##@siena07 siena07
siena07 <- function(x, batch = FALSE, verbose = FALSE, silent=FALSE,
	useCluster = FALSE, nbrNodes = 2,
	thetaValues = NULL,
	initC=TRUE,
	clusterString=rep("localhost", nbrNodes), tt=NULL,
	parallelTesting=FALSE, clusterIter=!x$maxlike,
	clusterType=c("PSOCK", "SOCK", "FORK", "MPI"),
	logLevelConsole='WARNING', logLevelFile='INFO',
	logIncludeLocation=F, cl=NULL,...)
{
	sienaSetupLogger(logLevelConsole=logLevelConsole,
		logLevelFile=logLevelFile,
		logBaseName=paste(x$projname, 0, sep='-'),
		logIncludeLocation=logIncludeLocation)

	exitfn <- function()
	{
		if (!is.batch())
		{
		  tcltk::tkdestroy(tkvars$tt)
		}
		## close the report file
		Report(closefiles=TRUE)
		RNGkind("default")
	}
	on.exit(exitfn())
	if (useCluster) {
		if (clusterType == "MPI") {
			nbrNodes <- max(Rmpi::mpi.comm.size(0) - 1, 1)
		}
	}

	# If the user is passing clusters through -cl- then change the
	# useCluster to TRUE, and assign the -nbrNodes- to number of nodes
	if (!useCluster & length(cl))
	{
		useCluster <- TRUE
		nbrNodes   <- length(cl)
	}

	time0 <-  proc.time()['elapsed']
	z <- NULL ## z is the object for all control information which may change.
	## x is designed to be readonly. Only z is returned.
	z$x <- x

	if (useCluster)
	{
		if (parallelTesting)
		{
			stop("cannot parallel test with multiple processes")
		}

		if (!length(cl))
		{
			clusterType <- match.arg(clusterType)
			if (.Platform$OS.type == "windows" && clusterType != "PSOCK")
			{
				stop("cannot use forking processes on Windows")
			}
		}
		# The possibility to use snow now has been dropped
		# because RSiena requires R >= 2.15.0
		# and snow is superseded.
		# if (getRversion() < "2.14.0")
		## fake this to recreate old results
		## if (TRUE)
		#{
		#	require(snow, warn.conflicts=FALSE)
		#	require(rlecuyer)
		#	clusterType <- "SOCK"
		#}
		#else
		#{
		## require(parallel)
		#}
		if (clusterIter)
		{
			x$firstg <- x$firstg * sqrt(nbrNodes)
			z$int <- nbrNodes
			z$int2 <- 1
		}
		else
		{
			z$int <- 1
			z$int2 <- nbrNodes
		}
	}
	else
	{
		z$int <- 1
		z$int2 <- 1
	}
	if (parallelTesting)
	{
		set.seed(1, kind='Wich')
		## randomseed2 is for second generator needed only for parallel testing
		randomseed2 <- .Random.seed
		## .Random.seed[2:4] <- as.integer(c(1,2,3))
		## randomseed2[2:4] <- as.integer(c(3,2,1))
		randomseed2[2:4] <- as.integer(c(1, 2, 3))
		seed <- 1
		newseed <- 1
		z$parallelTesting <- TRUE
	}
	else
	{
		randomseed2 <- NULL
		## x$randomSeed is the user seed, if any
		if (!is.null(x$randomSeed))
		{
			set.seed(x$randomSeed, kind="default")
			seed <- x$randomSeed
		}
		else
		{
			if (!nzchar(Sys.getenv("RSIENA_TESTING")))
			{
				if (exists(".Random.seed"))
				{
					rm(.Random.seed, pos=1)
				}
				newseed <- trunc(runif(1) * 1000000)
				set.seed(newseed)  ## get R to create a random number seed for me.
				seed <- NULL
			}
			else
			{
				newseed <- NULL
				seed <- NULL
			}
		}
	}
	z$randomseed2 <- randomseed2

	## set the global is.batch
	batchUse <- batch
	if (!batch)
	{
		if (!requireNamespace("tcltk", quietly = TRUE))
		{
				batchUse <- TRUE
				message("Package tcltk not available, forcing use of batch mode")		
		}
		else
		{
		if (.Platform$OS.type != "windows")
		{
			if (!capabilities("X11"))
			{
				batchUse <- TRUE
				message("No X11 device available, forcing use of batch mode")
			}
		}
		}
		if(nzchar(Sys.getenv("RSIENA_TESTING")))
		{
			silent <- TRUE
		}
	}
	is.batch(batchUse)

	## open the output file
	Report(openfiles=TRUE, projname=x$projname, verbose=verbose, silent=silent)
	z <- InitReports(z, seed, newseed)

	## reset the globals for interrupts
	NullChecks()

	## create the screen
	if (!is.batch())
	{
		tkvars <- siena07Gui(tt=tt)
		z$tkvars<- tkvars
		z$pb <- list(pb=tkvars$pb, pbval=0, pbmax=1)
	}
	else
	{
		z$pb <- list(pb=NULL, pbval=0, pbmax=1)
	}

	## create theta values for phase 3, if necessary
	if (is.null(thetaValues))
	{
		z$thetaValues <- NULL
		z$thetaFromFile <- FALSE
	}
	else
	{
		if (!x$simOnly)
		{
			cat('The thetaValues parameter was given\n')
			cat('but simOnly was not specified in the algorithm object.\n')
			cat('This is inconsistent.\n')
			stop('Inconsistent combination simOnly - thetaValues')
		}
		if (!is.matrix(thetaValues))
		{
			stop('thetaValues should be a matrix.')
		}
		z$thetaValues <- thetaValues
		z$thetaFromFile <- TRUE
	}
	z <- robmon(z, x, useCluster, nbrNodes, initC, clusterString,
		clusterIter, clusterType, cl, ...)
	time1 <-  proc.time()['elapsed']
	Report(c("Total computation time", round(time1 - time0, digits=2),
			"seconds.\n"), outf)

	if (useCluster)
	{
		# Only stop cluster if it wasn't provided by the user
		#	  if (!length(cl))
		#		  stopCluster(z$cl)

		## need to reset the random number type to the normal one
		assign(".Random.seed", z$oldRandomNumbers, pos=1)
	}

	class(z) <- "sienaFit"
	z$tkvars <- NULL
	z$pb <- NULL
	z
}

##@InitReports siena07 Print report
InitReports <- function(z, seed, newseed)
{
	Report("\n\n-----------------------------------\n", outf)
	Report("New Analysis started.\n", outf)
	Report(c("Date and time:", format(Sys.time(),"%d/%m/%Y %H:%M:%S")), outf)
	Report("\nNew results follow.\n", outf)
	Report("-----------------------------------\n", outf)
	rforgeRevision <-  packageDescription(pkgname,
		fields="Repository/R-Forge/Revision")
	if (is.na(rforgeRevision))
	{
		revision <- ""
	}
	else
	{
		revision <- paste(" R-forge revision: ", rforgeRevision, " ", sep="")
	}
	version <- packageDescription(pkgname, fields = "Version")
	Report(c(paste("\n", pkgname, " version ", sep = ""), version, " (",
		format(as.Date(packageDescription(pkgname, fields = "Date")), "%d %b %y"),
		")", revision, "\n\n"), sep = "", outf)
	if (z$x$simOnly)
	{
		Heading(1, outf, "Simulations.")
	}
	else
	{
		Heading(1, outf, "Estimation by stochastic approximation algorithm.")
	}
	if (is.null(seed))
	{
		Report("Random initialization of random number stream.\n", outf)
		if (!is.null(newseed))
		{
			Report(sprintf("Current random number seed is %d.\n", newseed), outf)
		}
	}
	else
	{
		Report(sprintf("Current random number seed is %d.\n", seed), outf)
	}
	z$revision <- revision
	z$version <- version
	z$startingDate <- date()
	z
}

##@AnnouncePhase siena07 Progress reporting
AnnouncePhase <- function(z, x, subphase=NULL)
{
	## require(tcltk)
	if (!is.batch())
	{
		tcltk::tkdelete(z$tkvars$phase, 0, "end")
		tcltk::tkinsert(z$tkvars$phase, 0, paste(" ", z$Phase))
		tcltk::tkdelete(z$tkvars$subphase, 0, "end")
		tcltk::tkdelete(z$tkvars$iteration, 0, "end")
		tcltk::tkinsert(z$tkvars$iteration, 0, format(0, width=6))
	}
	if (missing(subphase))
	{
		if (is.batch())
		{
			##Report(c("\nStart phase", z$Phase, "\n"), cf)
			Report(c("\nStart phase", z$Phase, "\n"))
		}
	}
	else
	{
		if (!is.batch())
		{
			tcltk::tkinsert(z$tkvars$subphase, 0, paste(" ", subphase))
		}
		else
		{
			##Report(c("\nStart phase ", z$Phase, ".", subphase, "\n"),
			##sep="", cf)
			Report(c("\nStart phase ", z$Phase, ".", subphase, "\n"), sep="")
		}
	}
	if (z$Phase == 0)
	{
		if (!is.batch())
		{
			tcltk::tkconfigure(z$tkvars$current, height=min(z$pp, 30))
			tcltk::tkconfigure(z$tkvars$deviation, height=min(z$pp, 30))
			tcltk::tkconfigure(z$tkvars$quasi, height=min(z$pp, 30))
		}
		n1pos <- z$n1 * (z$pp + 1)
		z$n2min0 <- 7 + z$pp
		z$n2min0 <- max(5, z$n2min0 / z$int)
		z$n2minimum<- rep(0, x$nsub)
		z$n2maximum<- rep(0, x$nsub)
		## 2.5198421 = 2^(4/3); this gives a gain parameter of order n^(-3/4) ##
		if (x$nsub > 0)
		{
			z$n2minimum[1] <-
				ifelse(is.null(x$n2start), trunc(z$n2min0 * 2.52), x$n2start)
			z$n2maximum[1] <- z$n2minimum[1] + 200
			if (x$nsub > 1)
			{
				for (i in 2:x$nsub)
				{
					z$n2minimum[i] <- trunc(z$n2minimum[i-1] * 2.52)
					z$n2maximum[i] <- z$n2minimum[i] + 200
				}
			}
		}
		z$n2partsum <- c(0, cumsum(z$n2maximum))
		n2sum <- sum(z$n2maximum)
		##Progress bar
		pbmax <- n1pos + n2sum + x$n3
		z$n1pos<- n1pos
		if (!x$maxlike && z$FinDiff.method)
			pbmax <- pbmax + x$n3 * z$pp
		z$pb$pbval <- 0
		z$pb <- createProgressBar(z$pb, maxvalue=pbmax)
		z$pb$pbmax <- pbmax
	}
	if (z$Phase==2)
	{
		propo <- z$n1pos + z$n2partsum[max(subphase,1)]
		if (propo> getProgressBar(z$pb))
			z$pb <-setProgressBar(z$pb,propo)
	}
	if (z$Phase ==3)
	{
		propo <- z$n1pos + z$n2partsum[x$nsub + 1]
		if (!z$AllUserFixed)
			z$pb <- setProgressBar(z$pb,propo)
		else
		{
			max <- x$n3
			z$pb <-createProgressBar(z$pb,max)
		}
	}
	z
}
##@roundfreq siena07 Prettify interval between progress reports
roundfreq <- function(w)
{
	vec1 <- c(1, 2, 3, 4, 31, 66, 101, 300, 500)
	vec2 <- c(1, 2, 3, 20, 50, 100, 200, 500)
	if (is.batch())
		w <- max(10,vec2[findInterval(w, vec1, all.inside=TRUE)])
	else
		w <- vec2[findInterval(w, vec1[1:7], all.inside=TRUE)]
	w
}

##@WriteOutTheta siena07 Progress reporting
WriteOutTheta <- function(z)
{
	if (!is.batch())
	{
		DisplayTheta(z)
	}
	else
	{
		Report(c("theta:", format(z$theta, digits=3), "\n"))
	}
	Report("Current parameter values:\n", cf)
	Report(format(z$theta), cf, fill=80)
}

##@DisplayThetaAutocor siena07 Progress reporting
DisplayThetaAutocor <- function(z)
{
	if (!is.batch())
	{
		DisplayTheta(z)
		tcltk::tkdelete(z$tkvars$quasi, "1.0", "end")
		tcltk::tkinsert(z$tkvars$quasi, "1.0", FormatString(z$pp, z$ac))
	}
	else
	{
		Report(c("theta", format(z$theta, digits=3),"\n"))
		Report(c("ac", format(z$ac, digits=3), "\n"))
	}

}
##@DisplayandWriteTheta siena07 Progress reporting
DisplayandWritetheta <- function(z)
{
	if (!is.batch())
	{
		DisplayTheta(z)
	}
	else
	{
		Report(c("theta", format(z$theta, digits=3), "\n"))
	}
}
##@DisplayTheta siena07 Progress reporting
DisplayTheta <- function(z)
{
	if (!is.batch())
	{
		tcltk::tkdelete(z$tkvars$current, "1.0", "end")
		tcltk::tkinsert(z$tkvars$current, "1.0", FormatString(z$pp, z$theta))
	}

}
##@FormatString siena07 Progress Reporting
FormatString <- function(pp, value)
{
	ppuse <- min(30, pp)
	nbrs <- format(1:ppuse)
	nch <- nchar(nbrs[1])
	formatstr <- paste("%", nch, "d.%", (13 - nch), ".4f\n", sep="",
		collapse="")
	paste(sprintf(formatstr, 1:ppuse, value[1:ppuse]), collapse="")
}
##@DisplayDeviations siena07 Progress reporting
DisplayDeviations <- function(z, fra)
{
	if (!is.batch())
	{
		tcltk::tkdelete(z$tkvars$deviations, "1.0", "end")
		tcltk::tkinsert(z$tkvars$deviations, "1.0", FormatString(z$pp, fra))
	}
}
##@DisplayIteration siena07 Progress reporting
DisplayIteration <- function(z)
{
	if (!is.batch())
	{
		tcltk::tkdelete(z$tkvars$iteration, 0, "end")
		tcltk::tkinsert(z$tkvars$iteration, 0, format(z$nit, width=6))
		tcltk::tcl("update")
	}
}
##@Root siena07 Safe square root for compatibility with siena3. Probably not necessary in R.
Root<- function(x)
{
	ifelse(abs(x) > 1e-36, sqrt(abs(x)), 1e-18)
}

##@getProgressBar siena07 Progress reporting
getProgressBar <- function(pb)
{
	if (is.batch())
		val <- pb$pbval
	else
		val <- as.numeric(tcltk::tclvalue(tcltk::tkcget(pb$pb, "-value")))
	val
}

##@setProgressBarProgress siena07 reporting
setProgressBar <- function(pb, val)
{
	if (is.batch())
	{
		pb$pbval <- val
	}
	else
	{
		tcltk::tkconfigure(pb$pb, value=val)
		tcltk::tcl("update")
	}
	pb
}
##@createProgressBar siena07 Progress reporting
createProgressBar <- function(pb, maxvalue)
{
	if (is.batch())
		pb$pbmax <- maxvalue
	else
		tcltk::tkconfigure(pb$pb, maximum=maxvalue)
	pb
}
##@tkErrorMessage Miscellaneous Not used
tkErrorMessage <- function()
{
	tcltk::tkmessageBox(geterrmessage(), icon="error")
}

##@errorHandler Miscellaneous Not used
errorHandler <- function()
{
	## opts <- options()
	if (!is.batch())
	{
		options(show.error.messages=FALSE)
		options(error=tkErrorMessage)
	}
}
